### ISSUE TITLE
Unable to `squeeze` array of lists

### ISSUE THREAD

**Issue by v0lat1le** [pydata/xarray#841]
Using python 2.7.10, xarray 0.7.2

```python
listarray = numpy.empty((1,), dtype=object)
listarray[0] = [1,2,3]  # note list type

ds = xarray.Dataset({'v2':(('b',), listarray)})
ds.squeeze()
```
this generates
`ValueError: dimensions () must have the same length as the number of data dimensions, ndim=1`

**Comment by jhamman** [comment]
This is an interesting corner case. Out of curiosity, what did you expect to get out of this? It seems we are converting listarray into a 2d array which is probably not the best choice here.

**Comment by v0lat1le** [comment]
Well, this works:
```python
listarray = numpy.empty((1,), dtype=object)
listarray[0] = {1:2, 3:4}  # anything but list works here

ds = xarray.Dataset({'v2':(('b',), listarray)})
ds.squeeze()
```
outputing
```
<xarray.Dataset>
Dimensions:  ()
Coordinates:
    b        int64 0
Data variables:
    v2       object {1: 2, 3: 4}
```

So I guess I'm expecting something like:
```
<xarray.Dataset>
Dimensions:  ()
Coordinates:
    b        int64 0
Data variables:
    v2       object [1, 2, 3]
```

In my use case I'd like to have an xarray.DataArray of lists of python objects to store some metadata per time-slice of the data

**Comment by MaximilianR** [comment]
Potentially related to the underlying cause of https://github.com/pydata/xarray/issues/847

**Comment by shoyer** [comment]
To provide a more direct example of the same underlying issue, `ds.sel(b=0)` also raises, for similar reasons.

This is hard to get right because NumPy array indexing can return scalars, instead of always returning NumPy arrays. When you index your `listarray`, for example, you get back a Python list, instead of a 0d array with dtype=object (which might or might not be what you want in this case).

I just wrote pushed a fix to my PR #867.
